bin/node/runtime/src/lib.rs

```diff
@@ -1368,23 +1368,31 @@ impl_runtime_apis! {
 
 	impl pallet_mmr::primitives::MmrApi<
 		Block,
-		mmr::Leaf,
 		mmr::Hash,
 	> for Runtime {
-		fn generate_proof(leaf_index: u64) -> Result<(mmr::Leaf, mmr::Proof<mmr::Hash>), mmr::Error> {
+		fn generate_proof(leaf_index: u64)
+			-> Result<(mmr::EncodableOpaqueLeaf, mmr::Proof<mmr::Hash>), mmr::Error>
+		{
 			Mmr::generate_proof(leaf_index)
+				.map(|(leaf, proof)| (mmr::EncodableOpaqueLeaf::from_leaf(&leaf), proof))
 		}
 
-		fn verify_proof(leaf: mmr::Leaf, proof: mmr::Proof<mmr::Hash>) -> Result<(), mmr::Error> {
+		fn verify_proof(leaf: mmr::EncodableOpaqueLeaf, proof: mmr::Proof<mmr::Hash>)
+			-> Result<(), mmr::Error>
+		{
+			let leaf: mmr::Leaf = leaf
+				.into_opaque_leaf()
+				.try_decode()
+				.ok_or(mmr::Error::Verify)?;
 			Mmr::verify_leaf(leaf, proof)
 		}
 
 		fn verify_proof_stateless(
```

Comment:
Perhaps `verify_ephemeral_proof` in order to emphasize, that it is not based on persistent data?

Reply:
I thought that `stateless` is enough emphasis :) "Ephemeral proof" sounds like a special kind of a proof to me, while here we just don't query the `state`, hence `stateless`.